### ISSUE TITLE
XMPP RFC supports

### ISSUE THREAD

**Issue by Neustradamus** [araditc/Artalk.Xmpp#3]
Please add support of last RFCs.

About XMPP RFCs:
- https://tools.ietf.org/html/rfc3920 -> https://tools.ietf.org/html/rfc6120 to https://tools.ietf.org/html/rfc7590
- https://tools.ietf.org/html/rfc3921 -> https://tools.ietf.org/html/rfc6121
- https://tools.ietf.org/html/rfc3922
- https://tools.ietf.org/html/rfc3923
- https://tools.ietf.org/html/rfc4622 -> https://tools.ietf.org/html/rfc5122
- https://tools.ietf.org/html/rfc6122 -> https://tools.ietf.org/html/rfc7622
- https://tools.ietf.org/html/rfc7711
- https://tools.ietf.org/html/rfc7712
- https://tools.ietf.org/html/rfc7395

Difference noted in RFCs:
- https://tools.ietf.org/html/rfc6120#appendix-D
- https://tools.ietf.org/html/rfc6121#appendix-E
- https://tools.ietf.org/html/rfc6122#appendix-C
- https://tools.ietf.org/html/rfc7622#appendix-A

Note:
- http://jabber.996255.n3.nabble.com/Changes-missing-from-RFC-6120-s-Differences-from-RFC-3920-td30486.html

RFC3920 -> RFC6120
- https://tools.ietf.org/tools/rfcdiff/rfcdiff.pyht?url1=https://www.ietf.org/rfc/rfc3920.txt&url2=https://www.ietf.org/rfc/rfc6120.txt

RFC3921 -> RFC6121
- https://tools.ietf.org/tools/rfcdiff/rfcdiff.pyht?url1=https://www.ietf.org/rfc/rfc3921.txt&url2=https://www.ietf.org/rfc/rfc6121.txt

Draft but Very Important:
- https://tools.ietf.org/html/draft-cridland-xmpp-session-01

**Comment by ammarheidari** [comment]
We welcome to all contributor. If you can develop, do it.

**Comment by ammarheidari** [comment]
Maintainer update: thank you for collecting the RFC references. The library historically started from RFC 3920/3921-era behavior, and not every RFC 6120/6121/6122/7622 update is fully reflected yet.

Artalk.Xmpp 2.0.0 has now been published for .NET 10. The modernization work removed several old platform/package issues, but full RFC alignment still needs a protocol-by-protocol pass. I am keeping this open as the umbrella tracking issue for RFC modernization.

The most useful next step is to split this into smaller implementation issues, for example:

- RFC 6120 stream and TLS/SASL behavior
- RFC 6121 session/roster/presence behavior
- RFC 7622 JID/stringprep handling
- RFC 7395 WebSocket binding

That will make the work reviewable and testable.

**Comment by ammarheidari** [comment]
Updated the supported feature documentation as part of phase 1.

Current `master` / NuGet `Artalk.Xmpp` version `2.1.0` now includes:

- .NET 10 target.
- STARTTLS and direct TLS.
- SASL `SCRAM-SHA-256`, `SCRAM-SHA-1`, `DIGEST-MD5`, and `PLAIN`.
- Optional legacy XMPP session establishment.
- IM, presence, roster, service discovery, ping, chat state, in-band registration, private XML storage, blocking, avatar/mood/tune/activity, and file-transfer support already documented in README.

BOSH, MUC, OMEMO, and the remaining SCRAM `-PLUS`/SHA variants are still separate follow-up items.

**Comment by ammarheidari** [comment]
Phase 2 update: basic XEP-0045 Multi-User Chat support is now implemented.

Current `master` / NuGet `Artalk.Xmpp` version `2.2.0` now adds:

- `JoinRoom(...)`
- `LeaveRoom(...)`
- `SendRoomMessage(...)`
- `RoomMessage` event for groupchat messages
- `RoomPresence` event for occupant presence

The README and XML examples were updated with a MUC sample. More advanced MUC features such as room configuration, invitations, moderation/admin flows, and bookmarks can be tracked as separate follow-up issues.

**Comment by ammarheidari** [comment]
Phase 3 update: connection hardening is now available in `Artalk.Xmpp` version `2.3.0`.

Added:

- `PingServer(...)` for XEP-0199 server pings.
- Optional automatic keepalive via `KeepAliveInterval` and `KeepAliveTimeout`.
- `KeepAliveFailed` event for ping timeout/failure visibility.
- Better stream failure signaling: reader failures now mark the client disconnected and raise both `Disconnected` and `Error`.
- Dispatcher event-handler exceptions are surfaced through `Error`.

This closes the long-running idle/stalled connection concern tracked in #22.

**Comment by ammarheidari** [comment]
Phase 4 update: additional SCRAM mechanisms are now implemented and published in `Artalk.Xmpp` version `2.4.0`.

This phase improves RFC/SASL alignment by adding:

- `SCRAM-SHA-224`
- `SCRAM-SHA-384`
- `SCRAM-SHA-512`
- `SCRAM-SHA3-512`

The README has been updated with the expanded SASL support list and the install command now points to `2.4.0`.

Validation:

- `dotnet test` passes: 15/15.
- `dotnet build -c Release` passes with 0 warnings and 0 errors.
- NuGet `Artalk.Xmpp` `2.4.0` was published and smoke-restored from nuget.org.

Commit: `922c9d6`

Remaining RFC modernization work still includes the SCRAM `-PLUS` channel-binding variants, BOSH/WebSocket transport work, JID/stringprep review, and larger protocol compliance passes.

**Comment by ammarheidari** [comment]
Phase 5 update: BOSH support is now implemented and published in `Artalk.Xmpp` version `2.5.0`.

This phase adds XMPP over BOSH support for HTTP-based connection managers:

- BOSH session open and restart support.
- Public `ArtalkXmppClient(Uri boshUrl, string hostname, string username, string password)` constructor.
- Public unauthenticated BOSH constructor for registration flows.
- `BoshUrl` property on client/core surfaces.
- Existing auth, bind, IQ, message, presence, roster, MUC, and keepalive logic reused over the BOSH transport.
- README and XML examples updated.

Validation:

- `dotnet test` passes: 18/18.
- `dotnet build -c Release --no-restore` passes with 0 warnings and 0 errors.
- NuGet `Artalk.Xmpp` `2.5.0` was published and smoke-restored from nuget.org.

Commit: `5d5b483`

This closes #23. Remaining larger RFC modernization work still includes SCRAM `-PLUS` channel binding, WebSocket binding, JID/stringprep review, and OMEMO.

**Comment by ammarheidari** [comment]
Phase 6 update: SCRAM-PLUS channel binding support is now implemented and published in `Artalk.Xmpp` version `2.6.0`.

Added:

- `SCRAM-SHA-1-PLUS`
- `SCRAM-SHA-224-PLUS`
- `SCRAM-SHA-256-PLUS`
- `SCRAM-SHA-384-PLUS`
- `SCRAM-SHA-512-PLUS`
- `SCRAM-SHA3-512-PLUS`
- `tls-server-end-point` channel binding for encrypted TCP XMPP streams

The client now prefers SCRAM-PLUS mechanisms when the server advertises them and TLS channel binding data is available.

Validation:

- `dotnet test` passes: 22/22.
- `dotnet build -c Release --no-restore` passes with 0 warnings and 0 errors.
- NuGet `Artalk.Xmpp` `2.6.0` was published and smoke-restored from nuget.org.

Commit: `1ee484e`

Remaining RFC/security follow-ups include `tls-unique` / `tls-exporter` if .NET exposes the required TLS material in the future, WebSocket binding, JID/stringprep review, Cisco Jabber diagnostics, and OMEMO.

**Comment by ammarheidari** [comment]
Phase 7 update: SASL `OAUTHBEARER` support is now implemented and published in `Artalk.Xmpp` version `2.7.0`.

Added:

- `OAUTHBEARER` SASL mechanism.
- `OAuthBearerToken` property on client/core surfaces.
- Token-aware SASL selection, so `OAUTHBEARER` is preferred when a bearer token is configured and advertised by the server.
- OAuth-only connection flow where `Username` and `OAuthBearerToken` can be set without a password.
- README and XML sample coverage.

Validation:

- `dotnet test` passes: 27/27.
- `dotnet build -c Release --no-restore` passes with 0 warnings and 0 errors.
- NuGet `Artalk.Xmpp` `2.7.0` was published and smoke-restored from nuget.org.

Commit: `17d2832`

This improves Cisco/Jabber-style compatibility for deployments that advertise standard `OAUTHBEARER`. `CISCO-VTG-TOKEN` remains a separate Cisco-specific item.

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.8.0` is published with the first XEP-0384 OMEMO foundation.

Added in this step:

- OMEMO device list and bundle PEP support
- Public models: `OmemoDeviceList`, `OmemoBundle`, `OmemoDeviceListChangedEventArgs`
- High-level client helpers for publishing/retrieving OMEMO device lists and bundles
- README and sample XML documentation updated
- Tests added and passing

Current validation:

- Release build succeeds for library and samples
- Test suite passes: 35 tests
- NuGet 2.8.0 is visible and restore-tested from nuget.org
- Commit: `abd8d3c`

Remaining OMEMO work is intentionally tracked separately from this foundation: encryption/decryption, session store, key/trust management, and encrypted message payload processing.

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.9.0` is now published with the next OMEMO phase.

This release updates OMEMO support to the current XEP-0384 namespace (`urn:xmpp:omemo:2`) and adds encrypted message envelope/payload helpers on top of the previous device list and bundle support.

Validation completed:

- Release build for library and samples succeeds with 0 warnings
- Test suite passes: 41 tests
- NuGet 2.9.0 is visible and restore-tested from nuget.org
- Commit: `f56dc8e`

Remaining OMEMO work: the full session/trust layer that creates and consumes recipient encrypted key material.

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.10.0` is now published with the OMEMO trust/session orchestration layer.

Added in this release:

- OMEMO trust-store abstraction and in-memory implementation
- Trust policy support with secure default `RequireTrusted`
- Pluggable `IOmemoSessionCipher` boundary for vetted X3DH / Double Ratchet implementations
- `OmemoSessionManager` for device discovery, bundle retrieval, trust checks, encryption envelope creation, and decrypt flow
- High-level client helpers for encrypting, sending, and decrypting OMEMO messages

Validation:

- Release build succeeds for library and samples with 0 warnings
- Test suite passes: 46 tests
- NuGet 2.10.0 is visible and restore-tested from nuget.org
- Commit: `d8e2d8a`

This keeps the package license-compatible while avoiding a hand-rolled Double Ratchet implementation in the core XMPP client.

**Comment by ammarheidari** [comment]
I am keeping this umbrella issue open for now.

A large part of the requested modernization has been completed across the 2.x releases, including .NET 10, modern SASL/SCRAM, direct TLS, BOSH, keepalive hardening, MUC basics, OAUTHBEARER, and the OMEMO XMPP-facing layers up through `2.10.0`.

However, this issue references a broad set of RFCs and transport updates. It is better to close it only after the remaining standards are either implemented or split into separate focused issues. In particular, RFC 7395 WebSocket binding and any remaining RFC-by-RFC compliance checks should be tracked explicitly rather than silently treating the umbrella as complete.

**Comment by ammarheidari** [comment]
RFC 7395 WebSocket binding is now implemented and published in `Artalk.Xmpp 2.11.0`.

Added:

- XMPP over WebSocket transport using the `xmpp` WebSocket subprotocol
- RFC 7395 framed stream open/close support with `urn:ietf:params:xml:ns:xmpp-framing`
- Standalone UTF-8 text frames for stanzas and stream elements
- Stream restart support after SASL, matching the existing TCP/BOSH negotiation flow
- Public `XmppTransportBinding.WebSocket` selector for URI-based client constructors
- `WebSocketUrl` properties on the core, IM, and high-level client layers
- README and XML examples updated
- Transport tests for subprotocol negotiation, open/restart frames, stanza send/receive, close frames, and URI validation

Validation:

- `dotnet test .\Artalk.XmppSoloution.sln -c Release` passed: 51 tests
- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore` passed with 0 warnings
- NuGet package `Artalk.Xmpp 2.11.0` published and verified with fresh restore from nuget.org
- Commit: `9ff68db`

This closes one of the major remaining RFC items from this umbrella issue. Next focused candidates are XEP-0454 OMEMO Media Sharing, a vetted `IOmemoSessionCipher` adapter/sample, or Cisco `CISCO-VTG-TOKEN` if protocol details become available.

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.12.0` is now published.

This release adds the next standards/compliance hardening step after WebSocket:

- RFC 7622-oriented JID parsing and normalization
- IDN-aware domainpart normalization through .NET IDN APIs
- Localpart normalization with case/width handling, RFC 7622 excluded-character checks, UTF-8 octet length enforcement, and control/surrogate validation
- Resourcepart opaque-string handling with NFC normalization and non-ASCII space mapping
- Correct handling for domain-only bare/full JIDs
- Additional JID unit coverage for IDN, resource separators, invalid localparts, non-ASCII spaces, supplementary Unicode, and 1023-octet limits
- CI workflow for restore/build/test/pack on `master` and PRs
- NuGet package hardening: SourceLink, symbol package, package tags, XML docs, deterministic build metadata
- README and sample XML updated for `2.12.0`

Validation:

- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore` passes with 0 warnings / 0 errors
- `dotnet test .\Artalk.XmppSoloution.sln -c Release --no-build` passes: 61/61
- `dotnet pack .\Artalk.Xmpp\Artalk.Xmpp.csproj -c Release --no-build` creates both `.nupkg` and `.snupkg`
- NuGet `Artalk.Xmpp 2.12.0` was published and fresh-restored from nuget.org

Commit: `b97a578`

I am keeping this umbrella issue open because the next concrete protocol item is still XEP-0454 OMEMO Media Sharing, and the Cisco-specific token path remains separate until protocol details are available.

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.13.0` is now published with XEP-0454 OMEMO Media Sharing helpers.

Added:

- `OmemoMediaUri` for strict `aesgcm://` URI creation and parsing
- AES-256-GCM media encryption/decryption with the authentication tag appended to the encrypted file
- HTTPS-only source URL validation when creating media URIs
- XEP-0454 encrypted length helper for HTTP upload slot sizing
- `OmemoMediaMessage` for strict media message bodies: an `aesgcm://` URL plus optional JPEG thumbnail data URI
- README and XML sample updates for send/receive flows
- Tests covering the official URI shape, strict parser rejection paths, AES-GCM round-trip, tamper detection, thumbnail parsing, and encrypted size calculation

Validation:

- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore` passes with 0 warnings / 0 errors
- `dotnet test .\Artalk.XmppSoloution.sln -c Release --no-build` passes: 79/79
- `dotnet pack .\Artalk.Xmpp\Artalk.Xmpp.csproj -c Release --no-build` creates `.nupkg` and `.snupkg`
- NuGet `Artalk.Xmpp 2.13.0` was published and fresh-restored from nuget.org

Commit: `61a9fb8`

This closes the main XEP-0454 media-sharing item that was left after the RFC 7395 and JID hardening work. I am keeping this umbrella issue open for any remaining RFC-by-RFC tracking or future split-out compliance issues.

**Comment by Neustradamus** [comment]
@ammarheidari: Thanks a lot for your recent work!

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.15.0` is now published with the XEP-0388 SASL2 foundation.

Added:

- XEP-0388 `<authentication xmlns=urn:xmpp:sasl:2>` stream-feature parsing
- SASL2 mechanism selection integrated with the existing SCRAM/OAUTHBEARER/PLAIN stack
- SASL2 `<authenticate/>`, `<initial-response/>`, `<challenge/>`, `<response/>`, `<success/>`, `<failure/>`, and `<additional-data/>` handling
- SCRAM server-signature verification from SASL2 `<additional-data/>`
- No post-success stream restart; the client waits for the authenticated `<stream:features/>` as required by XEP-0388
- Legacy RFC 6120 SASL remains the fallback path for older servers
- README/NuGet metadata updated for `2.15.0`

Validation:

- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore`: passed with 0 warnings / 0 errors
- `dotnet test .\Artalk.XmppSoloution.sln -c Release --no-build`: passed 87/87
- `dotnet pack .\Artalk.Xmpp\Artalk.Xmpp.csproj -c Release --no-build`: produced `.nupkg` and `.snupkg`
- NuGet `Artalk.Xmpp 2.15.0` was published and verified with a fresh restore from nuget.org

Commit: `9f49412`

I am keeping this umbrella open because the next SASL2-dependent items are still tracked separately: XEP-0474 (#31), XEP-0480 (#32), and XEP-0515 (#33).

**Comment by ammarheidari** [comment]
Status update: `Artalk.Xmpp 2.16.0` is now published with XEP-0480 SASL Upgrade Tasks support.

Added:

- SASL2 upgrade-task advertisement parsing via `urn:xmpp:sasl:upgrade:0`
- SCRAM upgrade requests in the SASL2 `<authenticate/>` element
- SASL2 `<continue/>` handling for requested SCRAM upgrade tasks
- `<next/>` task initiation and `<task-data><hash/></task-data>` SCRAM hash responses
- Salted-password derivation for the SCRAM upgrade tasks matching the supported SCRAM mechanisms
- README and NuGet metadata updated for `2.16.0`

Validation:

- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore`: passed with 0 warnings / 0 errors
- `dotnet test .\Artalk.XmppSoloution.sln -c Release --no-build`: passed 91/91
- `dotnet pack .\Artalk.Xmpp\Artalk.Xmpp.csproj -c Release --no-build`: produced `.nupkg` and `.snupkg`
- NuGet `Artalk.Xmpp 2.16.0` was published and verified with a fresh restore from nuget.org

Commit: `d99e873`

The remaining focused SASL follow-ups are now XEP-0474 (#31) and XEP-0515 (#33).

**Comment by ammarheidari** [comment]
Status update: the remaining SASL downgrade-protection follow-ups are now published.

`Artalk.Xmpp 2.17.0` added XEP-0474 SASL SCRAM Downgrade Protection:

- SSDP hash calculation over the active SASL mechanism list and XEP-0440 channel-binding type list
- SCRAM server-first-message `h` attribute verification
- Authentication failure before client proof on mismatch
- Works through both SASL1 and SASL2 SCRAM paths

`Artalk.Xmpp 2.18.0` added XEP-0515 TLS Channel-Binding Downgrade Protection for TCP TLS:

- TLS version encoding from the negotiated `SslStream` protocol
- SCRAM server-first-message `t` attribute verification
- Authentication failure before client proof on mismatch
- Tests covering TLS 1.2 / TLS 1.3 encoding and SCRAM match/mismatch behavior

Validation for the final 2.18.0 state:

- `dotnet build .\Artalk.XmppSoloution.sln -c Release --no-restore`: passed with 0 warnings / 0 errors
- `dotnet test .\Artalk.XmppSoloution.sln -c Release --no-build`: passed 97/97
- `dotnet pack .\Artalk.Xmpp\Artalk.Xmpp.csproj -c Release --no-build`: produced `.nupkg` and `.snupkg`
- NuGet `2.17.0` and `2.18.0` were both published and verified with fresh restores from nuget.org

Commits: `84defc9`, `fcdeda2`

At this point the linked SASL2/SCRAM/XEP-0440/XEP-0474/XEP-0480/XEP-0515 request chain is implemented, with #31, #32, and #33 closed.

**Comment by ammarheidari** [comment]
Closing this umbrella issue after the current modernization pass.

Since the .NET 10 restart, the broad RFC/protocol items from this thread have been split into concrete releases and implemented across the `2.x` line, including:

- .NET 10 packaging and NuGet release hardening
- STARTTLS/direct TLS and optional legacy XMPP session behavior
- modern SCRAM and SCRAM-PLUS SASL mechanisms
- `tls-server-end-point` channel binding, XEP-0440, XEP-0474, XEP-0480, and XEP-0515 SASL/SCRAM hardening
- RFC 7395 XMPP over WebSocket
- BOSH support
- RFC 7622-oriented JID parsing and normalization
- XEP-0045 MUC basics
- XEP-0199 ping / keepalive hardening
- XEP-0384 OMEMO foundation and XEP-0454 OMEMO media sharing helpers
- SASL `OAUTHBEARER` and Cisco `CISCO-VTG-TOKEN` compatibility

Latest published state:

- NuGet: `Artalk.Xmpp 2.19.0`
- Commit: `5f368e8`
- Build: passed with 0 warnings / 0 errors
- Tests: passed 101/101
- Vulnerable/outdated package checks: clean
- Fresh restore from nuget.org: passed

This issue has served well as the umbrella tracker, but keeping it open now makes future work harder to review. Any remaining RFC-by-RFC or XEP-by-XEP improvements should be opened as focused issues with a specific protocol behavior, test case, or interoperability target.
